langchain/callbacks/shared.py

```diff
@@ -0,0 +1,109 @@
+"""A shared CallbackManager."""
+
+import threading
+from typing import Any, Dict, List
+
+from langchain.callbacks.base import (
+    BaseCallbackHandler,
+    BaseCallbackManager,
+    CallbackManager,
+)
+from langchain.schema import AgentAction, LLMResult
+
+
+class Singleton:
+    """A thread-safe singleton class that can be inherited from."""
+
+    _instance = None
+    _lock = threading.Lock()
+
+    def __new__(cls) -> Any:
+        """Create a new shared instance of the class."""
+        if cls._instance is None:
+            with cls._lock:
+                # Another thread could have created the instance
+                # before we acquired the lock. So check that the
+                # instance is still nonexistent.
+                if not cls._instance:
+                    cls._instance = super().__new__(cls)
+        return cls._instance
+
+
+class SharedCallbackManager(Singleton, BaseCallbackManager):
+    """A thread-safe singleton CallbackManager."""
+
+    _callback_manager: CallbackManager = CallbackManager([])
+
+    def on_llm_start(
+        self, serialized: Dict[str, Any], prompts: List[str], **kwargs: Any
+    ) -> None:
+        """Run when LLM starts running."""
+        with self._lock:
```

Comment:
I wonder if this is worth the trouble? I get what it's for, but maybe a more useful pattern might be to have a SharedCallbackManager that always executes the handlers on the main thread (via the asyncio event loop). Thoughts?

Reply:
Yes, that could work. My preference would be to wait till tracing is released (a follow-up to this PR) before changing the way global callback manager works (execute all handlers on main thread); we have something working right now that will require some moderate-high changes to get it to work with asyncio. Happy to have this conversation in a couple of weeks

Reply:
yeah lets wait for now

Reply:
fyi this is what I meant, not actual async support, which I agree should be separate from this work

https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.call_soon_threadsafe